File: Engine.py
import pygame
import numpy as np
import os

from pygame.locals import *
from OpenGL.GL import *
from OpenGL.GLU import *
from win32api import GetSystemMetrics

import pdb
class Engine(object):
	"""
	Class concerning simulation rendering and user interaction
	"""
	def __init__(self, config, template):
		"""
		Creates OpenGL display and sets basic elements
		
		@param config:	Config object with the starting parameters
		@param template: A generic tile created by a Tile Generator object
		
		@field width: Window's width
		@field height: Window's height
		"""
		
		### Detects if fullscreen for an axis is active
		conf_width = config.getint('Display','WIDTH')
		conf_height = config.getint('Display','HEIGHT')
		
		if conf_width == -1:
			config['Display']['WIDTH'] = str(GetSystemMetrics(0)//template.x)
			
		if conf_height == -1:
			config['Display']['HEIGHT'] = str(GetSystemMetrics(1)//template.y)
			
		os.environ['SDL_VIDEO_WINDOW_POS'] = '0,0'
			
		### Screen size
		self.width = int(config.getint('Display','WIDTH') * template.x)
		self.height = int(config.getint('Display','HEIGHT') * template.y)
		
		### OpenGL screen initialization
		pygame.init()
		pygame.display.set_mode((self.width, self.height), OPENGL| DOUBLEBUF)
		pygame.display.set_caption("Langton's Ant")
		
		glViewport(0, 0, self.width, self.height)
		glMatrixMode(GL_PROJECTION)
		glLoadIdentity()
		gluOrtho2D(0.0, self.width, 0.0, self.height)
		glMatrixMode(GL_MODELVIEW)
		glLoadIdentity()
	
		glClearColor(0.0, 0.0, 0.0, 0.0)	
		glShadeModel(GL_SMOOTH)
		glClearDepth(1.0)
	
	def run(self, Anthill, ColorGen, TileGen):
		"""
		Runs simulation
		"""
		tile_IDs = TileGen.construct(ColorGen.palette)
		Map = dict({(0,0) : 0})	#Dictionary mapping the board  (Coord tuple : Tile_ID)
		
		speed = 50	#Iterations per frame
		pre_speed = 0
		
		run = True
		while(run):	
			### User Interaction
			for event in pygame.event.get():
				if event.type == QUIT:
					run = False
			
				if event.type == KEYDOWN:
					if event.key == K_ESCAPE:
						if ColorGen.write:
							ColorGen.save()		#Saves last palette
						run = False
					
					#Speed management
					elif event.key == K_w and speed <= 100:
						speed += 10
					elif event.key == K_s and speed >= 10:
						speed -= 10
						
					#Pause
					elif event.key == K_p:
						if pre_speed == 0:
							pre_speed = speed
							speed = 0
						else:
							speed = pre_speed
							pre_speed = 0
					
					#Shuffles current color palette
					elif event.key == K_r:
						np.random.shuffle(ColorGen.palette)
						tile_IDs = TileGen.reset(ColorGen.palette)
					
					#Generates new colorset
					elif event.key == K_g:
						ColorGen.random_palette(ColorGen.base)
						ColorGen.generate_colors()
						tile_IDs = TileGen.reset(ColorGen.palette)
						
			glClear(GL_COLOR_BUFFER_BIT)
			
			self.render_tiles(Map)
			
			for iter in range(speed):
				for ant in Anthill.ants:		#1 turn for each ant
					
					color = Map.get(tuple(ant.pos))
					if color == None:	#New tile is rendered
						color = 0
					
					Map.update({tuple(ant.pos) : (color+1) % (len(tile_IDs)+1)}) 
					
					ant.command(color)  #Move ant based on recovered rule
					
					ant.pos[0] %= self.width	#Wraps ant position
					ant.pos[1] %= self.height
				
			pygame.display.flip()
			
	def render_tiles(self, map):
		"""
		Paint tiles for the current frame
		
		@param map: Dict((x,y) : ID), Dictionary which contain the position of the tiles and their respective color
		"""
		
		for coord in map.keys():
			hue = map.get(coord)
			if hue != 0:
				glPushMatrix()
				glTranslate(coord[0],coord[1],0)
				glCallList(hue)
				glPopMatrix()
				
	
		

File: Engine.cpp
/***
 * @file Engine.cpp
 * OpenGL Window and rendering execution
 */
// Copyright 2019 by Julio Albornoz <gatalegath@protonmail.com>
// The License.txt file describes the conditions under which this software may be distributed.

#include "Engine.h"
#include "Camera.h"
#include <algorithm>
#include <glm/glm.hpp>

Engine::Engine(Config* config, Tile* tile, Monitor* monitor){
	/***
	* Creates OpenGL display and sets basic elements
	* 
	* @param config:	Config object with the starting parameters
	* @param template: A generic tile created by a Tile Generator object
	* 
	* @field width: Window's width
	* @field height: Window's height
	* """
	*/
	// Window Initialization
	/** If windows:{...} else if Linux: {...} **/
	
	window = monitor->GLFW_init();
	
	//OpenGL settings:
	monitor->OpenGL_init();
	
	// Grid Initialization
	if (config->adjust){
		config->GWidth = monitor->getScreenWidth();	//TODO: Handle this through a Param setter
		config->GHeight = monitor->getScreenHeight();
		Gwidth = config->GWidth;
		Gheight = config->GHeight;
		}	
		
	else{
		Gwidth = config->GWidth * tile->X;
		Gheight = config->GHeight * tile->Y;
		}
	}
	
void Engine::run(Colony* Anthill, Color_Generator* ColorGen, Tile_Generator* TileGen){
	
	TileGen->construct(ColorGen->palette);	//Compile OpenGL lists
	
	std::map<std::pair<int,int>,int> Map; //Move this to a Grid class, including the constructor commands from above
	Map[std::make_pair(0,0)] = 0;
	
	int vel[2] = {50,0};
	speed = vel;
	
	Camera visor = Camera(glm::vec3(0.0f, 0.0f, 3.0f),
						   glm::vec3(0.0f, 0.0, 0.0f),
						   glm::vec3(0.0f, 1.0f, 0.0f));
	
	//Start rendering
	while (!glfwWindowShouldClose(window)){
		
		//Move this to own function
		
		catchInput(ColorGen, TileGen, &visor);
		
		visor.updateVectors();
        glClear(GL_COLOR_BUFFER_BIT);
				  
		render_tiles(&Map);
		
		for (int iter = 0; iter < *speed ; iter++){
			for (auto &ant : *Anthill->ants){
				int color;
				std::pair<int, int> coord = std::make_pair(ant.pos[0], ant.pos[1]);
				if (Map.count(coord) == 0){
					color = 0;
					}
				else{
					color = Map[coord];
					}
				
				Map[coord] = (color+1) % (ColorGen->size);
				ant.command(color);
				
				ant.pos[0] = ((ant.pos[0] % Gwidth) + Gwidth) % Gwidth;
				ant.pos[1] = ((ant.pos[1] % Gheight) + Gheight) % Gheight;
				}
			}
        glfwSwapBuffers(window);
        glfwPollEvents();
		}
		
	if (ColorGen->write){	//Move this to ColorGen destructor
		ColorGen->save();
		}
	
	glfwTerminate();
	}
	
void Engine::catchInput(Color_Generator* ColorGen, Tile_Generator* TileGen, Camera* visor){
	//TODO: Improve this method
	
	//Exit
	if(glfwGetKey(window, GLFW_KEY_ESCAPE) == GLFW_PRESS){
		glfwSetWindowShouldClose(window, true);
		}
	
	//Speed Management
	else if (glfwGetKey(window, GLFW_KEY_UP) == GLFW_PRESS && *speed < 100){
		*speed += 10;
		}
		
	else if (glfwGetKey(window, GLFW_KEY_DOWN) == GLFW_PRESS && *speed > 10){
		*speed -= 10;
		}
	
	else if (glfwGetKey(window, GLFW_KEY_P) == GLFW_PRESS){
		(*speed == 0) ? speed-- : speed++;
		}
	
	//Shuffles current color palette
	else if (glfwGetKey(window, GLFW_KEY_R) == GLFW_PRESS){
		std::random_shuffle(ColorGen->palette->begin(), ColorGen->palette->end());
		TileGen->reset(ColorGen->palette);
		}
	
	//Generates new colorset
	else if (glfwGetKey(window, GLFW_KEY_G) == GLFW_PRESS){
		delete ColorGen->palette;
		ColorGen->random_palette(ColorGen->base);
		ColorGen->generate_colors();
		TileGen->reset(ColorGen->palette);
		}
	
	//Camera movement
	else if(glfwGetKey(window, GLFW_KEY_W) == GLFW_PRESS){
		visor->move(UP);
		}
	
	else if(glfwGetKey(window, GLFW_KEY_S) == GLFW_PRESS){
		visor->move(DOWN);
		}	
	else if(glfwGetKey(window, GLFW_KEY_D) == GLFW_PRESS){
		visor->move(RIGHT);
		}
	else if(glfwGetKey(window, GLFW_KEY_A) == GLFW_PRESS){
		visor->move(LEFT);
		}
	}

void Engine::render_tiles(std::map<std::pair<int,int>, int>* map){
	/***
	* Paint tiles for the current frame
	*
	* @param map: Dict((x,y) : ID), Dictionary which contain the position of 
	* the tiles and their respective color
	*/
	
	for (auto it = map->begin(); it != map->end(); it++){
		if (it->second != 0){
			glPushMatrix();
			glTranslatef(std::get<0>(it->first), std::get<1>(it->first),0);
			glCallList(it->second);
			glPopMatrix();
			}
		}
	}